### ISSUE TITLE
Example: Start server, run tests when gulp watch triggers

### ISSUE THREAD

**Issue by mikehaas763** [karma-runner/gulp-karma#16]
I've been trying to figure this out for a while now with no luck. Rather than letting Karma watch files and trigger new tests runs, I would like gulp.watch() to trigger a new test run. I've tried combinations of Karma server and runner.

1. Run a gulp task that starts a Karma server, and sets up a gulp watch.
2. When a file change triggers the gulp watch created in step 1, run tests using the existing Karma server started in step 1.

Is this possible with the existing API's? If so, can someone explain to me how and I'll submit a PR that shows such an example for others. :)

**Comment by pkozlowski-opensource** [comment]
@mikehaas763 the bitter truth is that Karma (as it stands today) and Gulp have some overlapping functionality, watching capability being one of them. There is no good way, that I would be aware of, of making gulp watches co-existing with gulp watches. The only thing that "works" is to have Karma watching output of the gulp build - which means that you end up with both system watching files, which slows down things etc.

In short: I'm not aware of any good system for having gulp.watch used with Karma. This is why this repo looks like it looks and this is the best setup I'm aware of. If you can think of a better setup, please do send a PR!

**Comment by mikehaas763** [comment]
That was depressing! If I ever end up getting karma runs triggered from gulp watches I'll submit a PR. Thanks. :)

**Comment by stramel** [comment]
@mikehaas763 @pkozlowski-opensource Not the best option obviously but you could have your gulp watch call a function that triggers karma to run. Downside is that karma doesn't stay running during that period of time.

**Comment by mikehaas763** [comment]
@stramel Ya that's pretty easy to do. My hope was to be able to start up the karma server, and let gulp watch trigger karma tests runs using the already running karma server.

**Comment by stramel** [comment]
@mikehaas763 I thought that was what you were going for. Let me know if you figure it out!

**Comment by goodfield** [comment]
@mikehaas763 after a lot of researching and try and error, this solution works for me: https://gist.github.com/goodfield/ce6f623de4d954c32f6a

**Comment by pkozlowski-opensource** [comment]
@goodfield this is very inefficient: 
* you've got both gulp and karma watching files
* you need to restart Karma on each file change.

This is something I would _not_ recommend to anyone.

**Comment by goodfield** [comment]
@pkozlowski-opensource not really:

1. the server get's started in background with --no-auto-watch (so karma does not watch the files ;-))
2. gulp just starts the karma runner, which executed the tests against the running karma server

**Comment by sanbor** [comment]
I just wanted to say thanks to @goodfield for sharing the snippet. It works well and indeed it's pretty efficient since it spawns the server only once in no watch mode and it only calls the runner when a file is modified.

I think this is a good solution since the `tdd` snippet provided in the example `gulpfie.js` in this repo never finishes.

**Comment by sanbor** [comment]
One thing to noticed with this approach is that the karma server is started outside a gulp task, which will launch karma server whatever task you run in gulp (even something like `gulp minify`). I tried spawning the process from another task but it doesn't work since the task will never finish.

I managed to run `karma.server` in detach mode and unload the server using `karma.stopper` but I don't think that it's worth the added complexity. I'll share the code in case somebody founds it useful, since this lets you run karma from Gulp without blocking a task.

```js
/**
 * Runs tests once
 */
gulp.task('test', function (done) {
    new karma.Server({
        configFile: __dirname + '/karma.conf.js',
        singleRun: true
    }, done).start();
});

/**
 * Spawns a karma server with file watch off
 */
gulp.task('karma:serve', function (done) {
    var server = new karma.Server({
        configFile: __dirname + '/karma.conf.js',
        autoWatch: false,
        detached: true
    }).start();
    log('karma server will be stopped when gulp is closed');

    // gulp process can terminate in different ways Ctrl+C http://stackoverflow.com/a/14032965
    var stopServer = function(options, err) {
        if (options.cleanup || options.exit) {
            karma.stopper.stop({port: 9876}, function(exitCode) {
                if (exitCode === 0) {
                    console.log('Server stop as initiated');
                }
                process.exit(exitCode);
            });
        } else if (err) {
            console.log(err.stack);
        }
    };

    // stop karma server when exit gulp
    process.on('exit', stopServer.bind(null,{cleanup:true}));

    //catches ctrl+c event
    process.on('SIGINT', stopServer.bind(null, {exit:true}));

    //catches uncaught exceptions
    process.on('uncaughtException', stopServer.bind(null, {exit:true}));

  // call done() so the task is marked as finished
  done();
});

/**
 * Invokes karma client to run.
 */
gulp.task('karma:run', function (done) {
    karma.runner.run({
        configFile: __dirname + '/karma.conf.js'
    }, function finished() {
        done();
    });
});

/**
 * If the karma server is not running innvokes karma:serve.
 * Invokes karma:run when a js file is changed.
 */
gulp.task('karma:watch', function () {
    watch('test/**/*.spec.js', function() {
        log('JS File changed. Running tests');
        gulp.start('karma:run');
    });
});

gulp.task('default', ['karma:serve', 'karma:watch']);
```
